src/application/controller/controller.c

```diff
@@ -1,2 +1,140 @@
+
+    // check queue creation
+    if (internal_state_queue == NULL || output_queue == NULL) {
+        init_status = W_FAILURE;
+        log_text("controller", "queue creation failed");
+    }
+
+    // initialize structs: non-placeholder values
+
+    // gain table initialization
+
+    // Log initialization status
+    if (init_status == W_SUCCESS) {
+        log_text("controller", "initialization successful");
+
+    } else {
+        log_text("controller", "initialization failed");
+    }
+
+    // return w_status_t state
+    return init_status;
+}
+
+/**
+ * Update controller with new state data - called by state estimation module
+ * @param new_state Latest state estimate from state estimation
+ * @return W_FAILURE if validation/queueing fails
+ */
+w_status_t controller_update_inputs(controller_input_t *new_state) {
+    if (xQueueOverwrite(internal_state_queue, new_state) == pdPASS) { // overwrite internal queue
+        log_text("controller", "latest output received, altitude %f", new_state->altitude);
+        return W_SUCCESS;
+    }
+    log_text("controller", "state overwritting fails");
+    return W_FAILURE;
+}
+
+/**
+ * Get most recent control output (cammanded angle and timestamp) - called by
+ * state estimation module
+ * @param output Pointer to store output -> type defined in controller
+ * @return W_FAILURE if no output available
+ */
+w_status_t controller_get_latest_output(controller_output_t *output) {
+    // return cammand angle + send timestamp
+
+    if (xQueuePeek(output_queue, output, timeout) == pdPASS) {
+        log_text("controller", "latest output returned");
+        return W_SUCCESS;
+    }
+    log_text("controller", "timeout: no output available");
+    return W_FAILURE;
+}
+
+/**
+ * Controller task function for RTOS
+ */
+void controller_task(void *argument) {
+    // get current flight phase
+    flight_phase_state_t current_phase = STATE_INIT;
+
+    while (true) {
+        // log phase transitions, specifics logged in flight phase
+        if (current_phase != flight_phase_get_state()) {
+            log_text("controller", "flight phase changed");
+        }
+
+#if current_phase == STATE_ACT_ALLOWED || current_phase == STATE_COAST // if is proper state
+
+        // wait for new state data (5ms timeout)
+        controller_input_t new_state_msg;
+        if (xQueueReceive(internal_state_queue, &new_state_msg, timeout) == pdPASS) {
+            controller_state.current_state = new_state_msg;
+            // validate data
+
+            // log data received
+            log_text("controller", "new state data received for controller computations");
+
+        } else {
+            log_text("controller", "no new state data received for controller computations");
+            controller_state.data_miss_counter++;
+
+            // if number of data misses exceed threshold, transition to safe mode
+        }
+
+        // calculate gains based on current conditions
```

Comment:
I am confused how this work is being divided, cause clearly most of the implementation (translation from MATLAB) is not complete here